railties/test/railties/http_request_test.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require "isolation/abstract_unit"
+require "stringio"
+require "rack/test"
+
+module RailtiesTest
+  class HttpRequestTest < ActiveSupport::TestCase
+    include ActiveSupport::Testing::Isolation
+    include Rack::Test::Methods
+
+    def setup
+      build_app
+
+      app_file "config/routes.rb", <<-RUBY
+        Rails.application.routes.draw do
+          post 'posts', :to => 'posts#create'
+        end
+      RUBY
+
+      controller "posts", <<-RUBY
+        class PostsController < ApplicationController
+          def create
+            render plain: request.raw_post
+          end
+        end
+      RUBY
+    end
+
+    def teardown
+      teardown_app
+    end
+
+    def boot_rails
+      require "#{app_path}/config/environment"
+    end
+
+    test "parses request raw_post correctly when Transfer-Encoding: chunked with not Content-Length value" do
+      boot_rails
+
+      header "Transfer-Encoding", "chunked"
+      header "Content-Length", ""
```

Comment:
What if the header is not set at all? I'm not sure the Rack spec even allows an empty String. It says:
>The CONTENT_LENGTH, if given, must consist of digits only.

Reply:
So I believe that Rack Test calculates and sends the `Content-Length` header by default, so this would be overriding this behaviour. The best course of action I believe would be to delete the header before sending a request, I will investigate on how to do that

Reply:
> According to to mozilla docs, that header should not be there at all?

https://httpwg.org/specs/rfc9112.html#field.transfer-encoding
> Early implementations of Transfer-Encoding would occasionally send both a chunked transfer coding for message framing and an estimated Content-Length header field for use by progress bars. This is why Transfer-Encoding is defined as overriding Content-Length, as opposed to them being mutually incompatible.

and
>A server MAY reject a request that contains both Content-Length and Transfer-Encoding or process such a request in accordance with the Transfer-Encoding alone.

But:
https://httpwg.org/specs/rfc9112.html#body.content-length
>A sender MUST NOT send a Content-Length header field in any message that contains a [Transfer-Encoding](https://httpwg.org/specs/rfc9112.html#field.transfer-encoding) header field.

Here, we're the server, so we may support having both, but we don't have to.

Reply:
I just understood what railties tests does under the hood: it creates a Rails app on the fly and starts it using the `Puma` web server. Since that fix was already done in Puma >= 5.6.5, the `Content-Length` header value is written there and then passed on to the Rails application, so we cannot properly reproduce the bug with that default setting.

I will try checking if I can run the test Rails app on a web server that does not handle that issue, that's what this fix is aiming for.

Reply:
I understand the issue, but with using `Rack::Test` there is not a good way of preventing the sending of a `Content-Length` header without overriding it with an empty string.

The value is set [here on Rack::MockRequest](https://github.com/rack/rack/blob/2-2-stable/lib/rack/mock.rb#L159), and this object is built when `post` is called from the test, so there is not much to do to prevent it besides setting a noop non-nil value on it before calling post 🤔 .

Reply:
Maybe it would be a nice suggestion to change `Rack::MockRequest` to only set "CONTENT_LENGTH" if there is no "Transfer-Encoding" header ?

Reply:
That code from `Rack::MockRequest` seems to imply you can use set the :input  to something else that's not a String. Not sure how doable it is to pass it from this test down to mock request, but if we can, crafting an input that doesn't respond to `size` should result in Rack not setting the content-length.